app/features/notes_generator/tools.py

```diff
@@ -0,0 +1,199 @@
+        orientation = self.args.orientation
+        pdf_filename = "generated_notes.pdf"
+        pagesize = landscape(LETTER) if orientation.lower() == "landscape" else portrait(LETTER)
+        logger.info(f"create PDF file from the notes, orientation =  {orientation}")
+        # Create document
+        doc = SimpleDocTemplate(pdf_filename, pagesize=pagesize, rightMargin=0.5*inch, leftMargin=0.5*inch)
+        styles = getSampleStyleSheet()
+
+        # Custom styles
+        title_style = ParagraphStyle(
+            "TitleStyle", parent=styles["Title"], fontSize=16, alignment=1, spaceAfter=12
+        )
+        concept_style = ParagraphStyle(
+            "ConceptStyle", parent=styles["Normal"], leftIndent=24, spaceAfter=6
+        )
+
+        # Build PDF content
+        content = []
+
+        # Title
+        content.append(Paragraph(notes_data["title"], title_style))
+        content.append(Spacer(1, 0.2 * inch))
+
+        # Summary
+        content.append(Paragraph(notes_data["summary"], styles["BodyText"]))
+        content.append(Spacer(1, 0.2 * inch))
+
+        # Numbered Major Key Concepts
+        for idx, major_concept in enumerate(notes_data["majorkeyconceptslist"], start=1):
+            # Display major concept title with numbering
+            major_concept_title = f"{idx}. {major_concept['majorconcept']}"
+            content.append(Paragraph(major_concept_title, styles["Heading2"]))
+
+            # Numbered sub-key concepts in Roman numerals
+            for sub_idx, key_concept in enumerate(major_concept["keyconceptdetails"], start=1):
+                sub_concept_title = f"{roman.toRoman(sub_idx)}. {key_concept['concept']}"
+                content.append(Paragraph(sub_concept_title, concept_style))
+
+                # Add each description as a paragraph below the concept title
+                for description in key_concept["conceptdescription"]:
+                    content.append(Paragraph(description, styles["BodyText"]))
+                    content.append(Spacer(1, 0.1 * inch))
+
+            # Add a spacer between major concepts
+            content.append(Spacer(1, 0.3 * inch))
+
+        # Generate PDF
+        try:
+            doc.build(content)
+        except Exception as e:
+            print(f"Error generating PDF: {str(e)}")
+
+        # Return the absolute path to the PDF file
+        full_path = os.path.abspath(pdf_filename)
+        return full_path
+    
+    def create_notes(self, documents: List[Document]):
+        logger.info(f"Creating the NOTES")
+
+        chain = self.compile(documents)
+
+         # Log the input parameters
+        input_parameters = (
+            f"Nb of columns: {self.args.nb_columns}, "
+            f"Topic: {self.args.topic}, "
+            f"Details: {self.args.details}"
+        )
+        logger.info(f"Input parameters: {input_parameters}")
+        attempt = 1
+        max_attempt = 6
+
+        while attempt < max_attempt:
```

Comment:
Make sure this doesn't take too long for generating the response.

Reply:
usually the response is generated from the first attempt